Add support for date & time handling (#435)
typst/typst: docs/src/reference/types.md

- returns: color

# Datetime

> Represents a datetime. Can be created by either specifying a custom datetime

Reviewer: I also write these kinds of comments all the time, but "Represents a datetime" is not really helpful. ;)

Maybe something like "Represents a date, a time, or a combination of both."?

using the [`datetime` function]($func/datetime) or getting the current date 
using the `today` function.

## Methods
### display()

Displays the datetime in a certain way. Depending on whether you have defined 

> the datetime as a date, time or datetime, the default format will be different.

Reviewer: ```suggestion
just a date, a time or both, the default format will be different.
```

> the datetime as a date, time or datetime, the default format will be different.

Reviewer: ```suggestion
just a date, a time or both, the default format will be different.
```

If you specified a date, it will be `[year]-[month]-[day]`. If you specified a
time, it will be `[hour]:[minute]:[second]`. In the case of a datetime, it will
be `[year]-[month]-[day] [hour]:[minute]:[second]`.

> The format of a datetime is specified by providing a so-called
> _components_ with a specified number of _modifiers_. A component represents a 

Reviewer: ```suggestion
The format of a datetime is specified by providing
_components_ with a specified number of _modifiers_. A component represents a 
```

  numerical representation of the week starts with 0 or 1.
* `hour`: Displays the hour of the date.
  * `padding`: Can be either `zero`, `space` or `none`. Specifies how the hour
  is padded.
  * `repr`: Can be either `24` or `12`. Changes whether the hour is displayed in
  the 24-hour or 12-hour format.
* `period`: The AM/PM part of the hour
  * `case`: Can be `lower` to display it in lower case and `upper` to display it
    in upper case.
* `minute`: Displays the minute of the date.
  * `padding`: Can be either `zero`, `space` or `none`. Specifies how the minute
    is padded.
* `second`: Displays the second of the date.
  * `padding`: Can be either `zero`, `space` or `none`. Specifies how the second
    is padded.

Keep in mind that not always all components can be used. For example, if
you create a new datetime with `#datetime(year: 2023, month: 10, day: 13)`, it
will be stored as a plain date internally, meaning that you cannot use 
components such as `hour` or `minute`, which would only work on datetimes 
that have a specified time.

- pattern: string (positional)
  The format used to display the datetime.
- returns: string

> ```example

Reviewer: The example should be way further up. It also contains too much text and long lines. If you look at the other examples in the documentation, they contain typically just a few lines and most fit into max. ~40 chars so that they can display side-to-side. Maybe you can make it a bit more compact. Don't worry about the exact number of chars or length, I will check how it looks and adapt if necessary.

Reviewer: I just noticed that all of this is in the `display()` method's docs. I think methods can't have examples right now, but in any case, I would most stuff up into the top-level docs. Maybe first the introductory text, then the example, and then a subheading `# Format` followed by the detailed explanation of components.

Author: Tried my best to shorten it a bit. :) And rearranged the entries.

#let date = datetime(year: 2020, month: 10, day: 4)

Default representation: #date.display() \
Custom representation: #date.display("Last two digits of year: [year padding:zero repr:last_two] Month: [month padding:zero], Weekday: [weekday repr:monday one_indexed:true] ([weekday repr:long]), Day: [day padding:none]")

#let time = datetime(hour: 18, minute: 2, second: 23)

Default representation: #time.display() \
Custom representation: #time.display("Hour: [hour repr:12] ([period]), Minute: [minute padding:none], Second: [second]")
```

### year()

> 

Reviewer: ```suggestion
```
This is of course a matter of taste, but I would prefer no blank lines after headings. :)
